### ISSUE TITLE
mutate: add mutation operator for constant replacement

### ISSUE THREAD

**Issue by christoffer-nylen** [joakim-brannstrom/dextool#1240]
It might be interesting to add an operator for "constant replacement".

For definitions of constant replacement mutation operators, see p61 in 
["Design of Mutant Operators for the C Programming Language"](https://www.researchgate.net/publication/2590658_Design_Of_Mutant_Operators_For_The_C_Programming_Language) :
>![image](https://user-images.githubusercontent.com/5559216/102375888-cc9c1e00-3fa1-11eb-8a8f-d9d135125a49.png)
>![image](https://user-images.githubusercontent.com/5559216/102376116-0d943280-3fa2-11eb-82c2-f836cb3273db.png)
>![image](https://user-images.githubusercontent.com/5559216/102376509-78de0480-3fa2-11eb-93a0-a006f0f7ecb3.png)

Example from ["What It Would Take to Use Mutation Testing in Industry - A Study at Facebook"](https://arxiv.org/abs/2010.13464) :
>![image](https://user-images.githubusercontent.com/5559216/102234247-db1a0500-3ecf-11eb-91ba-496c4edce3c0.png)

Example from "An Empirical Evaluation of Mutation Testing for Improving the Test Quality of Safety-Critical Software":
>![image](https://user-images.githubusercontent.com/5559216/102234812-6f846780-3ed0-11eb-8714-e597406df391.png)
>![image](https://user-images.githubusercontent.com/5559216/102235002-a195c980-3ed0-11eb-8854-702ca256b7c3.png)

**Comment by christoffer-nylen** [comment]
TL;DR; I think this example is still not good enough. It seems that other mutant operators will subsume the need for constant replacement (at least in this example).

Background:
The user wants to check that a test case has been implemented correctly. The user (me) is lazy and hopes that mutation testing will solve this problem! The user executes the recommended mutation operators (lcr, lcrb, sdl, uoi, dcr) + Checks all mutants reported as *unique*. The idea is that, if lucky, at least one of the unique mutants will provide a strong evidence that the test case has been implemented correctly.

However, when inspecting the unique mutants, the user realizes that they have all been killed due to segmentation faults in initialization code... Duh! Thinking that the test case should indeed have been more awesome than this, the users suspects that mutation points might be missing!!

The user decides to check if there could be something fishy going on in the test double configurations and begins to inspect the function call-chains that leads to the execution of the mocks. Using a debugger, the user confirms a call to `myCopyStuff` to result in the execution of the mock.

The user thinks that there should be interesting mutants along this path (that might be non-uniquely killed by the test or perhaps surviving).

Problem: When looking at the mutation testing report, it does not seem to be any mutant that temper with the invocation of `myCopyStuff`, could this be one of the reasons why "unique mutants are missing" ?

```
$ cat foo.c
307: unsigned short int retCode;
...
377: retCode = myCopyStuff(source, destination, data);
...
383: if (retCode != OK) {
...
```

The user realizes that there was an attempt by the tool to create an SDL mutant at the line:

```
$ dextool mutate generate --id 1482
$ cat foo.c
307: unsigned short int retCode;
...
377: ;
...
383: if (retCode != OK) {
...
```

But this mutant resulted in a compile error:

```
$ gcc -Wall -Werror -c foo.c
...
cc1: warnings being treated as errors
foo.c:383: error: 'retCode' may be used uninitialized in this function
```

The `-Wall` and `-Werror` options forces this kind of (undefined behavior) mutant to be killed by the compiler, which is a good thing! But, at first glance it makes the user (me) think that a mutants might be missing here! Which brings me back to this suggestion of adding an operator for constant replacement.

**Constant Replacement examples:**

```
307: unsigned short int retCode;
...
377: retCode = 0;
...
383: if (retCode != OK) {
...
```

```
307: unsigned short int retCode;
...
377: retCode = 65535;
...
383: if (retCode != OK) {
...
```
Pros:
* Does not get killed by the compiler.
* Covers the "critical path" that the user is investigating.

Cons:
* Expensive to maintain (schemata implementation)
* A lot more mutants -> Slower mutation testing.
* This might be redundant?! Consider LCR at line 383 `if (retCode != OK) {` -> `if(0) {` and `if (retCode != OK) {` -> `if(1) {` !

Conclusion:
* I think a more convincing example is still needed.

**Comment by joakim-brannstrom** [comment]
Closing the issue. CR is now implemented and the exit code is saved. The exit code can then be seen in the test case specific report.
